File: plugin/suite/roster/orchestration/src/glob_containment.py
"""Exact language containment for the selector's glob dialect.

Answers one question: is every path matched by glob `I` also matched by at
least one glob in a set `E`? That is language containment, `L(I) ⊆ ⋃L(E)`,
and it is *decidable* here -- the dialect (`**/`, `**`, `*`, `?`, literals)
describes regular languages, so the question has an exact yes/no answer.

It is used to detect a routing rule whose `exclude_paths` fully shadow one of
its own `paths` globs (issue #162), where the answer must be trustworthy in
both directions: a false "shadowed" verdict fails CI on a correct
routing.json, and a false "alive" verdict is the silent coverage loss the
check exists to catch.

An earlier version of that check sampled synthesized probe paths and reported
a glob when every probe was excluded. Sampling cannot support that verdict:
the finding is a *universal* claim, so an incomplete sample makes it strictly
easier to satisfy, which is the false-accusation direction. Concretely,
`paths: ["roster/**"]` with `exclude_paths: ["**/*.txt"]` was reported as
fully shadowed because every synthesized probe happened to end in `.txt`.
This module replaces that guess with a decision procedure.

**Method.** Each glob becomes an NFA over a finite *abstract* alphabet: the
distinct literal characters appearing in the globs under comparison, plus
`/`, plus one sentinel standing for every other character, plus `\n`. That
abstraction is exact rather than approximate, because the dialect only ever
tests a character for being a specific literal, for being `/`, or for being
anything -- it can never distinguish two characters that appear in no
pattern. `\n` earns its own symbol because the matcher is inconsistent about
it: `**` compiles to `.`, which excludes `\n` without `re.DOTALL`, while
`*`/`?` compile to `[^/]`, which includes it.

The one caveat is case folding: literals are folded with `str.lower()`, which
agrees with `re.IGNORECASE` across ASCII but not across all of Unicode (`ſ`
matches `s` under `IGNORECASE` while `"ſ".lower()` is unchanged; `K` matches
`k` but is not a pattern literal). Every such divergence keeps characters
*apart* that the matcher would merge, which can only cost a missed finding --
never a false one -- and no routing pattern is non-ASCII.

The answer is then computed by searching the product of the include NFA with
the determinized union of the exclude NFAs for a reachable state that accepts
the include and no exclude. Such a state yields a witness path: a concrete
counterexample, returned by `contains_with_witness` so that a NOT_CONTAINED
verdict can be *checked* against `glob_to_regex` rather than trusted. No
witness anywhere in the (finite) product means containment holds.

The search is bounded by `_MAX_PRODUCT_STATES`. Exhausting it returns
`UNDETERMINED` -- never a verdict -- so a caller that reports only on
`CONTAINED` cannot be made to accuse falsely by a pathological pattern.
"""

from __future__ import annotations

from collections import deque
from typing import Iterable

from routing import (
    GLOB_DOUBLESTAR,
    GLOB_DOUBLESTAR_SLASH,
    GLOB_LITERAL,
    GLOB_QUESTION,
    GLOB_STAR,
    iter_glob_tokens,
)

CONTAINED = "contained"
NOT_CONTAINED = "not-contained"
UNDETERMINED = "undetermined"

# The sentinel standing for "any character that appears in no pattern under
# comparison". One is sufficient: the dialect cannot tell two such characters
# apart, so a language that contains one contains them all. Abstract symbols
# are opaque tokens rather than characters, so this is deliberately a string
# no single-character pattern literal can ever equal.
_OTHER = "\0other"
_SEPARATOR = "/"
# `\n` needs its own symbol because the matcher treats it inconsistently:
# `glob_to_regex` compiles `**` to `.` (which excludes `\n` without
# re.DOTALL) but `*`/`?` to `[^/]` (which includes it). Folding `\n` into
# _OTHER would model `**` as consuming it, making `contains("foo/*",
# ["foo/**"])` wrongly CONTAINED -- a false accusation, since `foo/a\nb`
# really does match the include and not the exclude.
_NEWLINE = "\n"

# Bound on explored product states. Reaching it yields UNDETERMINED rather
# than a verdict. routing.json's real patterns explore a few dozen.
_MAX_PRODUCT_STATES = 50_000


class _Nfa:
    """An epsilon-NFA whose transitions are predicates over abstract symbols.

    States are allocated as needed rather than one per token, because `**/`
    needs an extra one (see below). State `0` is the start; `self.accept` is
    the single accepting state.
    """

    __slots__ = ("accept", "accepting", "epsilon", "loops", "moves", "_next")

    def __init__(self, pattern: str) -> None:
        # moves[state] -> list of (predicate_kind, literal_or_None, target)
        self.moves: dict[int, list[tuple[str, str | None, int]]] = {}
        self.loops: dict[int, str] = {}
        self.epsilon: dict[int, list[int]] = {}
        self._next = 1
        state = 0
        for kind, text in iter_glob_tokens(pattern):
            target = self._allocate()
            if kind == GLOB_LITERAL:
                self.moves.setdefault(state, []).append((GLOB_LITERAL, text.lower(), target))
            elif kind == GLOB_QUESTION:
                self.moves.setdefault(state, []).append((GLOB_QUESTION, None, target))
            elif kind == GLOB_STAR:
                # `[^/]*`: consume non-separators in place, then move on. The
                # skip stays available after looping, which is correct -- the
                # loop and the exit are the same alternative.
                self.loops[state] = GLOB_STAR
                self.epsilon.setdefault(state, []).append(target)
            elif kind == GLOB_DOUBLESTAR:
                # `.*`: same shape, over every symbol.
                self.loops[state] = GLOB_DOUBLESTAR
                self.epsilon.setdefault(state, []).append(target)
            elif kind == GLOB_DOUBLESTAR_SLASH:
                # `(?:.*/)?` is a *choice* between nothing and "anything, then
                # `/`" -- and the second branch, once entered, is committed to
                # ending on a separator. Modelling it as a self-loop plus an
                # epsilon skip on one state conflates the two: after the loop
                # consumed a character the skip would still be reachable, so
                # `**/a` would wrongly accept `.a`. The committed branch gets
                # its own state.
                consumed = self._allocate()
                self.epsilon.setdefault(state, []).extend((consumed, target))
                self.loops[consumed] = GLOB_DOUBLESTAR
                self.moves.setdefault(consumed, []).append((GLOB_LITERAL, _SEPARATOR, target))
            state = target
        self.accept = state
        # `glob_to_regex` anchors with `$`, not `\Z`, and `$` also matches
        # immediately before a single trailing newline -- so `glob_to_regex("a")`
        # really does match `"a\n"`. Modelling acceptance as the final state
        # alone would make the include's language smaller than the matcher's,
        # which is again the false-accusation direction.
        trailing_newline = self._allocate()
        self.moves.setdefault(state, []).append((GLOB_LITERAL, _NEWLINE, trailing_newline))
        self.accepting = frozenset({state, trailing_newline})

    def _allocate(self) -> int:
        state = self._next
        self._next += 1
        return state

    def closure(self, states: Iterable[int]) -> frozenset[int]:
        pending = deque(states)
        seen = set(pending)
        while pending:
            state = pending.popleft()
            for target in self.epsilon.get(state, ()):
                if target not in seen:
                    seen.add(target)
                    pending.append(target)
        return frozenset(seen)

    def step(self, states: frozenset[int], symbol: str) -> frozenset[int]:
        nxt: set[int] = set()
        for state in states:
            loop = self.loops.get(state)
            # `**` compiles to `.`, which excludes `\n`; `*` compiles to
            # `[^/]`, which includes it. The asymmetry is the matcher's, and
            # modelling it is what makes this decision exact.
            if loop == GLOB_DOUBLESTAR and symbol != _NEWLINE:
                nxt.add(state)
            elif loop == GLOB_STAR and symbol != _SEPARATOR:
                nxt.add(state)
            for kind, literal, target in self.moves.get(state, ()):
                if kind == GLOB_LITERAL and symbol == literal:
                    nxt.add(target)
                elif kind == GLOB_QUESTION and symbol != _SEPARATOR:
                    nxt.add(target)
        return self.closure(nxt)


def _alphabet(patterns: Iterable[str]) -> list[str]:
    """The finite abstract alphabet sufficient to decide these patterns."""
    symbols = {_SEPARATOR, _OTHER, _NEWLINE}
    for pattern in patterns:
        for kind, text in iter_glob_tokens(pattern):
            if kind == GLOB_LITERAL:
                symbols.add(text.lower())
    return sorted(symbols)


def _sample_path(pattern: str) -> str | None:
    """A concrete path matching `pattern`, used when there are no excludes."""
    verdict, witness = contains_with_witness(pattern, ["\0impossible-exclude"])
    return witness if verdict == NOT_CONTAINED else None


def _concrete(symbol: str, literals: set[str]) -> str:
    """Render an abstract symbol as a real character, for witness paths."""
    if symbol != _OTHER:
        return symbol
    for candidate in "zqxjkvwy0123456789":
        if candidate not in literals:
            return candidate
    return "￿"


def contains(include: str, excludes: Iterable[str]) -> str:
    """Return CONTAINED, NOT_CONTAINED, or UNDETERMINED for `L(include) ⊆ ⋃L(excludes)`."""
    return contains_with_witness(include, excludes)[0]


def contains_with_witness(include: str, excludes: Iterable[str]) -> tuple[str, str | None]:
    """As `contains`, but also return a concrete counterexample path.

    On NOT_CONTAINED the second element is a path that `include` matches and
    no exclude does -- independently checkable against `glob_to_regex`, which
    is what lets a test verify this verdict rather than take it on trust. It
    is `None` for CONTAINED and UNDETERMINED, where no such path exists or
    none was found.

    Matching is case-insensitive, mirroring `glob_to_regex`'s `re.IGNORECASE`.
    """
    exclude_patterns = list(excludes)
    if not exclude_patterns:
        return NOT_CONTAINED, _sample_path(include)
    include_nfa = _Nfa(include)
    exclude_nfas = [_Nfa(pattern) for pattern in exclude_patterns]
    alphabet = _alphabet([include, *exclude_patterns])

    literals = {
        symbol for symbol in _alphabet([include, *exclude_patterns]) if symbol not in (_OTHER,)
    }

    start = (
        include_nfa.closure([0]),
        tuple(nfa.closure([0]) for nfa in exclude_nfas),
    )
    seen = {start}
    parents: dict[tuple, tuple[tuple, str] | None] = {start: None}
    pending = deque([start])
    while pending:
        if len(seen) > _MAX_PRODUCT_STATES:
            return UNDETERMINED, None
        current = pending.popleft()
        include_states, exclude_states = current
        if (include_nfa.accepting & include_states) and not any(
            nfa.accepting & states for nfa, states in zip(exclude_nfas, exclude_states)
        ):
            # A path matched by the include and by no exclude: proof of life.
            symbols: list[str] = []
            walk = current
            while parents[walk] is not None:
                walk, symbol = parents[walk]  # type: ignore[misc]
                symbols.append(symbol)
            witness = "".join(_concrete(symbol, literals) for symbol in reversed(symbols))
            return NOT_CONTAINED, witness
        for symbol in alphabet:
            successor = (
                include_nfa.step(include_states, symbol),
                tuple(nfa.step(states, symbol) for nfa, states in zip(exclude_nfas, exclude_states)),
            )
            if not successor[0]:
                # The include can no longer match; this branch proves nothing.
                continue
            if successor not in seen:
                seen.add(successor)
                parents[successor] = (current, symbol)
                pending.append(successor)
    return CONTAINED, None

File: internal/orchestration/glob_containment.go
// glob_containment.go ports roster/orchestration/src/glob_containment.py:
// exact language containment for the selector's glob dialect.
//
// Answers one question: is every path matched by glob `include` also
// matched by at least one glob in `excludes`? That is language containment,
// L(include) ⊆ ⋃L(excludes), and it is decidable here -- the dialect
// (**, *, ?, literals) describes regular languages, so the question has an
// exact yes/no answer.
//
// It is used to detect a routing rule whose exclude_paths fully shadow one
// of its own paths globs -- where the answer must be trustworthy in both
// directions: a false "shadowed" verdict fails a check on a correct
// routing.json, and a false "alive" verdict is the silent coverage loss the
// check exists to catch. Sampling synthesized probe paths cannot support
// that verdict (it is a *universal* claim -- an incomplete sample makes it
// strictly easier to satisfy, which is the false-accusation direction), so
// this is a decision procedure, not a heuristic.
//
// The dialect modelled here is selector.GlobToRegex's -- the function that
// compiles every route path glob, and therefore the only matcher whose
// answers this check is about. Getting that wrong does not make the check
// stricter or looser in a useful way; it makes it a check on a language
// nothing matches in.
//
// Two properties of that dialect, both of which this file previously got
// backwards, and both found by the brute-force differential rather than by
// reading:
//
//   - Case: selector.GlobToRegex sets `(?i)`. Literals are folded here to
//     match. The earlier note claiming otherwise cited a private copy of
//     globToRegex that lived in this file and had no non-test callers; that
//     copy has been deleted in favour of calling the real matcher.
//   - `**`: the compilation depends on what follows. `**/` becomes
//     `(?:.*/)?` -- nothing, or anything ending on a separator -- while a
//     `**` with no `/` after it becomes `.*`. Treating both as the former
//     understates a trailing `**`, and an understated include is easier to
//     contain, which is the false-accusation direction.
//
// One property it got right, worth keeping written down: Go's regexp `$`
// (no `(?m)`) matches only the absolute end of text, unlike Python's `$`,
// which also matches just before a single trailing newline. No extra
// acceptance state is needed here, unlike the Python original.
//
// Method: each glob becomes an NFA over a finite abstract alphabet: the
// distinct literal bytes appearing in the globs under comparison, plus `/`,
// plus one sentinel standing for every other byte, plus `\n` (which needs
// its own symbol because `[^/]` includes it while `.` excludes it -- the
// same asymmetry the Python original documents). The answer is computed by
// searching the product of the include NFA with the exclude NFAs for a
// reachable state that accepts the include and no exclude; such a state
// yields a concrete witness path. No witness anywhere in the (finite,
// bounded) product means containment holds.
//
// A `[...]` character class anywhere in a pattern under comparison forces
// Undetermined rather than attempting to model arbitrary regexp
// character-class semantics -- routing.json's real patterns never use one,
// and a false verdict would be strictly worse than declining to decide.
package orchestration

import (
	"sort"
	"strings"
)

// Containment verdicts.
const (
	Contained    = "contained"
	NotContained = "not-contained"
	Undetermined = "undetermined"
)

// maxContainmentProductStates bounds explored product states during the
// containment search. Reaching it yields Undetermined rather than a
// verdict, so a caller that reports only on Contained cannot be made to
// accuse falsely by a pathological pattern. routing.json's real patterns
// explore a few dozen.
const maxContainmentProductStates = 50_000

const (
	symOther = "\x00other" // Sentinel: any byte that appears in no pattern under comparison.
	symSep   = "/"
	symNL    = "\n"
)

type globTokenKind int

const (
	globLiteral globTokenKind = iota
	globQuestion
	globStar
	// globDoubleStar is `**` with a `/` after it, which the matcher compiles
	// to `(?:.*/)?` -- nothing, or anything ending on a separator.
	globDoubleStar
	// globDoubleStarAny is `**` with no `/` after it, which the matcher
	// compiles to `.*` -- anything at all, separators included, not required
	// to end on one.
	//
	// The two were one token. Modelling a trailing `**` as `(?:.*/)?` makes
	// the pattern's language narrower than it really is, and a narrower
	// include is easier to contain: `**` came back contained by `**/`, which
	// matches only the empty string and paths ending in `/`.
	globDoubleStarAny
)

type globToken struct {
	kind globTokenKind
	lit  string // one-byte string; valid when kind == globLiteral
}

// tokenizeGlobForContainment re-derives the token stream that
// route_matching.go's globToRegex would compile glob into, without building
// the regexp.Regexp itself. Returns ok=false if the pattern contains a `[`
// (a character class) -- see package doc.
func tokenizeGlobForContainment(glob string) ([]globToken, bool) {
	var tokens []globToken
	for i := 0; i < len(glob); i++ {
		c := glob[i]
		switch c {
		case '*':
			if i+1 < len(glob) && glob[i+1] == '*' {
				i++
				// Which `**` this is depends on what follows it, so the
				// separator has to be looked at before the token is emitted.
				// The previous version consumed the `/` and emitted the same
				// token either way, which is how the distinction was lost.
				if i+1 < len(glob) && glob[i+1] == '/' {
					i++
					tokens = append(tokens, globToken{kind: globDoubleStar})
				} else {
					tokens = append(tokens, globToken{kind: globDoubleStarAny})
				}
			} else {
				tokens = append(tokens, globToken{kind: globStar})
			}
		case '?':
			tokens = append(tokens, globToken{kind: globQuestion})
		case '[':
			return nil, false
		default:
			// Regex-special characters (.+^$()|{}) are escaped by
			// globToRegex to match literally, same as any other byte --
			// the token stream doesn't distinguish them.
			//
			// Folded to lower case because the matcher this models --
			// selector.GlobToRegex, which compiles every route path glob --
			// sets (?i). Keeping case here made the analyzer answer a question
			// nobody asks: it reported `**/README.md` as not contained by
			// `**/readme.md` while the live matcher has the exclude swallowing
			// the include whole. Folding at the token stream keeps the
			// alphabet, the transitions and the witness consistent, and a
			// lower-case witness is a valid witness under a case-insensitive
			// matcher.
			tokens = append(tokens, globToken{kind: globLiteral,
				lit: strings.ToLower(string(c))})
		}
	}
	return tokens, true
}

type globMove struct {
	kind   globTokenKind // globLiteral or globQuestion
	lit    string        // valid when kind == globLiteral
	target int
}

// containmentNFA is an epsilon-NFA whose transitions are predicates over
// abstract symbols, mirroring the shape of the Python original's `_Nfa`.
type containmentNFA struct {
	moves   map[int][]globMove
	loops   map[int]globTokenKind // globStar or globDoubleStar
	epsilon map[int][]int
	accept  int
	next    int
}

func newContainmentNFA(tokens []globToken) *containmentNFA {
	n := &containmentNFA{
		moves:   map[int][]globMove{},
		loops:   map[int]globTokenKind{},
		epsilon: map[int][]int{},
		next:    1,
	}
	state := 0
	for _, tok := range tokens {
		target := n.allocate()
		switch tok.kind {
		case globLiteral:
			n.moves[state] = append(n.moves[state], globMove{kind: globLiteral, lit: tok.lit, target: target})
		case globQuestion:
			n.moves[state] = append(n.moves[state], globMove{kind: globQuestion, target: target})
		case globStar:
			// [^/]*: consume non-separators in place, then move on.
			n.loops[state] = globStar
			n.epsilon[state] = append(n.epsilon[state], target)
		case globDoubleStar:
			// (?:.*/)?: a choice between nothing and "anything but \n,
			// then /" -- the second branch, once entered, is committed to
			// ending on a separator, so it gets its own state (a self-loop
			// plus a skip on one state would wrongly let the loop continue
			// after a literal '/' was already consumed as the "then /"
			// step).
			consumed := n.allocate()
			n.epsilon[state] = append(n.epsilon[state], consumed, target)
			n.loops[consumed] = globDoubleStar
			n.moves[consumed] = append(n.moves[consumed], globMove{kind: globLiteral, lit: symSep, target: target})
		case globDoubleStarAny:
			// `.*`: any run of symbols but a newline -- separators included --
			// and no obligation to end on one. Same loop predicate as the
			// branch above, without the commitment to a final separator.
			n.loops[state] = globDoubleStar
			n.epsilon[state] = append(n.epsilon[state], target)
		}
		state = target
	}
	n.accept = state
	return n
}

func (n *containmentNFA) allocate() int {
	s := n.next
	n.next++
	return s
}

func (n *containmentNFA) closureSet(states map[int]bool) (string, map[int]bool) {
	seen := map[int]bool{}
	pending := []int{}
	for s := range states {
		seen[s] = true
		pending = append(pending, s)
	}
	for len(pending) > 0 {
		s := pending[0]
		pending = pending[1:]
		for _, t := range n.epsilon[s] {
			if !seen[t] {
				seen[t] = true
				pending = append(pending, t)
			}
		}
	}
	return stateSetKey(seen), seen
}

func (n *containmentNFA) step(states map[int]bool, symbol string) (string, map[int]bool) {
	next := map[int]bool{}
	for s := range states {
		loop, hasLoop := n.loops[s]
		if hasLoop {
			if loop == globDoubleStar && symbol != symNL {
				next[s] = true
			} else if loop == globStar && symbol != symSep {
				next[s] = true
			}
		}
		for _, mv := range n.moves[s] {
			switch mv.kind {
			case globLiteral:
				if symbol == mv.lit {
					next[mv.target] = true
				}
			case globQuestion:
				if symbol != symSep {
					next[mv.target] = true
				}
			}
		}
	}
	return n.closureSet(next)
}

func (n *containmentNFA) accepts(states map[int]bool) bool {
	return states[n.accept]
}

func stateSetKey(states map[int]bool) string {
	ids := make([]int, 0, len(states))
	for s := range states {
		ids = append(ids, s)
	}
	sort.Ints(ids)
	key := ""
	for _, id := range ids {
		key += string(rune(id)) + ","
	}
	return key
}

// containmentAlphabet is the finite abstract alphabet sufficient to decide
// these patterns: the distinct literal bytes in patterns, plus / and \n and
// the OTHER sentinel.
func containmentAlphabet(patternsTokens [][]globToken) []string {
	symbols := map[string]bool{symSep: true, symOther: true, symNL: true}
	for _, tokens := range patternsTokens {
		for _, tok := range tokens {
			if tok.kind == globLiteral {
				symbols[tok.lit] = true
			}
		}
	}
	out := make([]string, 0, len(symbols))
	for s := range symbols {
		out = append(out, s)
	}
	sort.Strings(out)
	return out
}

func concreteSymbol(symbol string, literals map[string]bool) string {
	if symbol != symOther {
		return symbol
	}
	for _, candidate := range "zqxjkvwy0123456789" {
		c := string(candidate)
		if !literals[c] {
			return c
		}
	}
	return "￿"
}

// GlobContains returns Contained, NotContained, or Undetermined for
// L(include) ⊆ ⋃L(excludes), under route_matching.go's actual glob dialect.
func GlobContains(include string, excludes []string) string {
	verdict, _ := GlobContainsWithWitness(include, excludes)
	return verdict
}

// GlobContainsWithWitness is GlobContains, but also returns a concrete
// counterexample path. On NotContained, the witness is a path that include
// matches and no exclude does -- independently checkable against
// route_matching.go's globToRegex. It is "" for Contained and Undetermined.
func GlobContainsWithWitness(include string, excludes []string) (string, string) {
	includeTokens, ok := tokenizeGlobForContainment(include)
	if !ok {
		return Undetermined, ""
	}
	if len(excludes) == 0 {
		return notContainedSample(include, includeTokens)
	}

	excludeTokensList := make([][]globToken, 0, len(excludes))
	for _, pattern := range excludes {
		tokens, ok := tokenizeGlobForContainment(pattern)
		if !ok {
			return Undetermined, ""
		}
		excludeTokensList = append(excludeTokensList, tokens)
	}

	includeNFA := newContainmentNFA(includeTokens)
	excludeNFAs := make([]*containmentNFA, len(excludeTokensList))
	for i, tokens := range excludeTokensList {
		excludeNFAs[i] = newContainmentNFA(tokens)
	}

	allTokens := append([][]globToken{includeTokens}, excludeTokensList...)
	alphabet := containmentAlphabet(allTokens)
	literals := map[string]bool{}
	for _, s := range alphabet {
		if s != symOther {
			literals[s] = true
		}
	}

	type productState struct {
		include  map[int]bool
		excludes []map[int]bool
	}
	type productKey struct {
		include  string
		excludes string
	}

	_, startInclude := includeNFA.closureSet(map[int]bool{0: true})
	startExcludes := make([]map[int]bool, len(excludeNFAs))
	startExcludeKeys := make([]string, len(excludeNFAs))
	for i, nfa := range excludeNFAs {
		_, s := nfa.closureSet(map[int]bool{0: true})
		startExcludes[i] = s
		startExcludeKeys[i] = stateSetKey(s)
	}

	makeKey := func(includeStates map[int]bool, excludeStates []map[int]bool) productKey {
		exKeys := ""
		for _, s := range excludeStates {
			exKeys += stateSetKey(s) + "|"
		}
		return productKey{include: stateSetKey(includeStates), excludes: exKeys}
	}

	start := productState{include: startInclude, excludes: startExcludes}
	startKey := makeKey(start.include, start.excludes)

	seen := map[productKey]bool{startKey: true}
	type parentEdge struct {
		key    productKey
		symbol string
	}
	parents := map[productKey]*parentEdge{startKey: nil}
	pending := []productState{start}
	pendingKeys := []productKey{startKey}

	for len(pending) > 0 {
		if len(seen) > maxContainmentProductStates {
			return Undetermined, ""
		}
		current := pending[0]
		currentKey := pendingKeys[0]
		pending = pending[1:]
		pendingKeys = pendingKeys[1:]

		if includeNFA.accepts(current.include) {
			anyExcludeAccepts := false
			for i, nfa := range excludeNFAs {
				if nfa.accepts(current.excludes[i]) {
					anyExcludeAccepts = true
					break
				}
			}
			if !anyExcludeAccepts {
				// Reconstruct the witness path by walking parent links.
				var symbols []string
				walkKey := currentKey
				for parents[walkKey] != nil {
					edge := parents[walkKey]
					symbols = append(symbols, edge.symbol)
					walkKey = edge.key
				}
				witness := ""
				for i := len(symbols) - 1; i >= 0; i-- {
					witness += concreteSymbol(symbols[i], literals)
				}
				return NotContained, witness
			}
		}

		for _, symbol := range alphabet {
			_, nextInclude := includeNFA.step(current.include, symbol)
			if len(nextInclude) == 0 {
				// The include can no longer match; this branch proves nothing.
				continue
			}
			nextExcludes := make([]map[int]bool, len(excludeNFAs))
			for i, nfa := range excludeNFAs {
				_, ns := nfa.step(current.excludes[i], symbol)
				nextExcludes[i] = ns
			}
			nextKey := makeKey(nextInclude, nextExcludes)
			if !seen[nextKey] {
				seen[nextKey] = true
				parents[nextKey] = &parentEdge{key: currentKey, symbol: symbol}
				pending = append(pending, productState{include: nextInclude, excludes: nextExcludes})
				pendingKeys = append(pendingKeys, nextKey)
			}
		}
	}
	return Contained, ""
}

// notContainedSample finds a concrete path matching include, used when
// there are no excludes at all (trivially NotContained).
func notContainedSample(include string, includeTokens []globToken) (string, string) {
	// An exclude pattern built only from literal bytes that cannot equal
	// any witness this search would produce -- the same trick the Python
	// original uses ("\0impossible-exclude") to reuse the general search
	// for the "no excludes at all" case instead of a separate code path.
	verdict, witness := GlobContainsWithWitness(include, []string{"\x00impossible-exclude"})
	if verdict == NotContained {
		return NotContained, witness
	}
	return NotContained, ""
}

// CheckRouteExcludeShadowing checks whether route's exclude_paths fully
// shadow any one of its own paths globs -- the routing.json correctness bug
// this file exists to catch. Returns, per paths entry: Contained (fully
// shadowed -- a real bug), NotContained (fine), or Undetermined.
func CheckRouteExcludeShadowing(route Route) map[string]string {
	results := make(map[string]string, len(route.Paths))
	for _, path := range route.Paths {
		results[path] = GlobContains(path, route.ExcludePaths)
	}
	return results
}
